Apps/Src/ReadCarCAN.c

```diff
@@ -67,98 +93,145 @@ static inline void chargingEnable(void) {
     // if nothing is precharging and we need to turn the array on, turn it on
     shouldRestartArray = !restartingArray && (Contactors_Get(ARRAY_CONTACTOR)==OFF) && (Contactors_Get(ARRAY_PRECHARGE)==OFF);
 
-    // start precharge for array 
+    // wait for precharge for array 
     if(shouldRestartArray){
 
-         if (Contactors_Set(ARRAY_PRECHARGE, ON) == SUCCESS) {    // only run through precharge sequence if we successfully start precharge
-            restartingArray = true;
+            // Wait to make sure precharge is finished and then restart array
+            OSTmrStart(&prechargeDlyTimer, &err);
             
-            //Turn on the array restart thread
-            OSTaskCreate(
-                &arrayTCB,
-                "Array Restarter",
-                &ArrayRestart,
-                NULL,
-                TASK_READ_CAR_CAN_PRIO,
-                arraySTK,
-                DEFAULT_STACK_SIZE/10,
-                DEFAULT_STACK_SIZE,
-                0,
-                1,
-                NULL,
-                OS_OPT_TASK_STK_CLR,
-                &err
-            );
-            assertOSError(OS_READ_CAN_LOC,err);   
         }
-    }
+    
 
     OSMutexPost(&arrayRestartMutex,
                 OS_OPT_NONE,
                 &err);
     assertOSError(OS_READ_CAN_LOC,err);
 }
 
+/**
+ * @brief Callback function for the precharge delay timer. Turns off precharge and restarts the array.
+ * @param p_tmr pointer to the timer that calls this function, passed by timer
+ * @param p_arg pointer to the argument passed by timer
+ * 
+*/
+static void arrayRestart(void *p_tmr, void *p_arg){
```

Comment:
This function can probably be simplified. `restartingArray` is set to false no matter what, so it can be done at the end and not in the `if` statement. Setting the Contactors and updating the display can be put in the `if` block together, with no `else` necessary.